src/main/java/uk/gov/hmcts/reform/sandl/snlevents/actions/hearing/AssignSessionsToHearingAction.java

```diff
@@ -0,0 +1,166 @@
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.HearingPart;
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.Session;
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.UserTransactionData;
+import uk.gov.hmcts.reform.sandl.snlevents.model.request.HearingSessionRelationship;
+import uk.gov.hmcts.reform.sandl.snlevents.model.request.SessionAssignmentData;
+import uk.gov.hmcts.reform.sandl.snlevents.repository.db.HearingRepository;
+import uk.gov.hmcts.reform.sandl.snlevents.repository.db.SessionRepository;
+import uk.gov.hmcts.reform.sandl.snlevents.service.RulesService;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import javax.persistence.EntityManager;
+
+public class AssignSessionsToHearingAction extends Action implements RulesProcessable {
+
+    protected final EntityManager entityManager;
+    protected HearingSessionRelationship relationship;
+    protected UUID hearingId;
+    protected Hearing hearing;
+    protected List<Session> targetSessions;
+    protected List<UUID> targetSessionsIds;
+    protected List<String> previousHearingParts;
+    protected String previousHearing;
+
+    protected HearingRepository hearingRepository;
+    protected SessionRepository sessionRepository;
+    protected Hearing savedHearing;
+
+    public AssignSessionsToHearingAction(UUID hearingId,
+                                         HearingSessionRelationship relationship,
+                                         HearingRepository hearingRepository,
+                                         SessionRepository sessionRepository,
+                                         EntityManager entityManager,
+                                         ObjectMapper objectMapper) {
+        this.relationship = relationship;
+        this.hearingId = hearingId;
+        this.hearingRepository = hearingRepository;
+        this.sessionRepository = sessionRepository;
+        this.objectMapper = objectMapper;
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public void getAndValidateEntities() {
+        hearing = hearingRepository.findOne(hearingId);
+        if (hearing == null || hearing.getHearingParts() == null || hearing.getHearingParts().isEmpty()) {
+            throw new SnlEventsException("Hearing cannot be null!");
+        }
+
+        if (relationship.getSessionsData() == null) {
+            throw new SnlEventsException("SessionsData cannot be null!");
+        }
+        targetSessionsIds = relationship.getSessionsData().stream()
+            .map(SessionAssignmentData::getSessionId)
+            .collect(Collectors.toList());
+
+        targetSessions = sessionRepository.findSessionByIdIn(targetSessionsIds);
+        if (targetSessions == null || targetSessions.isEmpty()) {
+            throw new SnlEventsException("Target sessions cannot be null!");
```

Comment:
I think you should be iterating through all list and check if any of the session is null. Or maybe nulls are not returned from findSessionByIdIn method? In that case we have to make sure that the amount of targetSessionId's is the same as the amount of targetSessions

Reply:
Might be that it should also be checking if hearing parts and target sessions number matches.
WIll add both.
And yes, empty/null rows are not returned.